### PR TITLE
[REVIEW] Concatenate MultiIndex from groupby results for dask-cudf

### PR DESCRIPTION
**Pull request by thomcom**
Resolves https://github.com/rapidsai/dask-cudf/issues/125

### CODE REVIEW

**Comment by thomcom** [comment]
Resolved your issues @kkraus14. CI seems to be having its own.

**Comment by raydouglass** [comment]
rerun tests

**Comment by thomcom** [comment]
rerun tests

**Comment by raydouglass** [comment]
rerun tests

File: python/cudf/dataframe/series.py [hunk 1]
```diff
@@ -611,7 +611,12 @@ def dtype(self):
     def _concat(cls, objs, axis=0, index=True):
         # Concatenate index if not provided
         if index is True:
-            index = Index._concat([o.index for o in objs])
+            from cudf.dataframe.multiindex import MultiIndex
+            if isinstance(objs[0].index, MultiIndex):
+                index = MultiIndex._concat([o.index.copy(
+                    deep=False) for o in objs])
```
Comment (kkraus14) [inline python/cudf/dataframe/series.py]: Why do we need shallow copies of the objects feeding into the concat call?

File: python/cudf/dataframe/multiindex.py [hunk 2]
```diff
@@ -336,6 +339,53 @@ def __getitem__(self, index):
     def _values(self):
         return list([i for i in self])
 
+    @classmethod
+    def _concat(cls, objs):
+        _need_codes = False
+        from cudf import DataFrame
+        from cudf import MultiIndex
+        for idx, o in enumerate(objs):
```
Comment (kkraus14) [inline python/cudf/dataframe/multiindex.py]: can replace this whole loop with:
```python
_need_codes = not all([hasattr(o, '_source_data') for o in objs])
```

File: python/cudf/dataframe/multiindex.py [hunk 3]
```diff
@@ -336,6 +339,53 @@ def __getitem__(self, index):
...
+            if not hasattr(o, '_source_data'):
+                _need_codes = True
+        if _need_codes:
```
Comment (kkraus14) [inline python/cudf/dataframe/multiindex.py]: Unless there's a compelling reason, lets just make this raise `NotImplementedError` for now until we can handle this more efficiently or avoid it entirely 😄

File: python/cudf/dataframe/series.py (hunk 1 above)
Comment (thomcom) [reply python/cudf/dataframe/series.py]: The _needs_codes logic modified the .codes dataframe columns, which broke my algorithm by recomputing on intermediate results.

Comment (thomcom) [reply python/cudf/dataframe/series.py]: Should not be necessary now.

File: python/cudf/dataframe/multiindex.py [hunk 4]
```diff
@@ -336,6 +339,21 @@ def __getitem__(self, index):
     def _values(self):
         return list([i for i in self])
 
+    @classmethod
+    def _concat(cls, objs):
+        _need_codes = False
```
Comment (kkraus14) [inline python/cudf/dataframe/multiindex.py]: Don't need this anymore:
```suggestion
```

**Pull request merged**